File: android/app/src/main/java/com/readypackets/mobile/core/MobileApiClient.kt
package com.readypackets.mobile.core

import com.readypackets.mobile.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.KSerializer
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer
import okhttp3.FormBody
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.util.UUID

class MobileApiClient(private val store: SecureTokenStore, private val client: OkHttpClient = OkHttpClient()) {
    @PublishedApi internal val json = Json { ignoreUnknownKeys = true; explicitNulls = false }
    private val refreshMutex = Mutex()
    private val portal = BuildConfig.PORTAL_BASE_URL.trimEnd('/')

    suspend inline fun <reified T> get(path: String): Result<T> = request(path, "GET", null, null, serializer<T>())
    suspend inline fun <reified T, reified B> post(path: String, body: B): Result<T> = request(path, "POST", json.encodeToString(body), UUID.randomUUID().toString(), serializer<T>())
    suspend inline fun <reified T, reified B> put(path: String, body: B): Result<T> = request(path, "PUT", json.encodeToString(body), UUID.randomUUID().toString(), serializer<T>())
    suspend inline fun <reified T> delete(path: String): Result<T> = request(path, "DELETE", null, UUID.randomUUID().toString(), serializer<T>())
    suspend fun getBytes(path: String): Result<ByteArray> = binaryRequest(path, "GET", null, null)

    suspend fun upload(path: String, files: List<File>, fields: Map<String, String>): Result<CustomerFilesPage> {
        val multipart = MultipartBody.Builder().setType(MultipartBody.FORM).apply {
            fields.forEach { (key, value) -> addFormDataPart(key, value) }
            files.take(5).forEach { file -> addFormDataPart("files", file.name, file.readBytes().toRequestBody(mimeType(file).toMediaType())) }
        }.build()
        return binaryRequest(path, "POST", multipart, UUID.randomUUID().toString())
            .mapCatching { json.decodeFromString<CustomerFilesPage>(it.decodeToString()) }
    }

    suspend fun exchangeCode(code: String, verifier: String): Result<TokenSet> = tokenRequest(FormBody.Builder().add("grant_type", "authorization_code").add("client_id", "readypackets-native").add("code", code).add("code_verifier", verifier).add("redirect_uri", BuildConfig.OAUTH_REDIRECT_URI).build())

    suspend fun revoke() {
        val token = store.refreshToken ?: return
        withContext(Dispatchers.IO) { client.newCall(Request.Builder().url("$portal/api/mobile/v1/revoke").post(FormBody.Builder().add("token", token).add("client_id", "readypackets-native").build()).build()).execute().close() }
    }

    @PublishedApi internal suspend fun <T> request(path: String, method: String, body: String?, idempotency: String?, responseSerializer: KSerializer<T>, retry: Boolean = true): Result<T> = withContext(Dispatchers.IO) {
        val token = validAccessToken().getOrElse { return@withContext Result.failure(it) }
        val builder = Request.Builder().url("$portal/api/mobile/v1$path").header("Authorization", "Bearer $token").header("Accept", "application/json")
        if (idempotency != null) builder.header("Idempotency-Key", idempotency)
        when (method) { "POST" -> builder.post((body ?: "").toRequestBody("application/json".toMediaType())); "PUT" -> builder.put((body ?: "").toRequestBody("application/json".toMediaType())); "DELETE" -> builder.delete(); else -> builder.get() }
        client.newCall(builder.build()).execute().use { response ->
            val payload = response.body?.string().orEmpty()
            if (response.code == 401 && retry) { refresh().getOrElse { return@withContext Result.failure(it) }; return@withContext request(path, method, body, idempotency, responseSerializer, retry = false) }
            if (!response.isSuccessful) return@withContext Result.failure(problem(payload, response.code))
            runCatching { json.decodeFromString(responseSerializer, payload) }
        }
    }

    private suspend fun binaryRequest(path: String, method: String, body: RequestBody?, idempotency: String?, retry: Boolean = true): Result<ByteArray> = withContext(Dispatchers.IO) {
        val token = validAccessToken().getOrElse { return@withContext Result.failure(it) }
        val builder = Request.Builder().url("$portal/api/mobile/v1$path").header("Authorization", "Bearer $token").header("Accept", "application/json")
        if (idempotency != null) builder.header("Idempotency-Key", idempotency)
        when (method) { "POST" -> builder.post(body ?: ByteArray(0).toRequestBody()); "PUT" -> builder.put(body ?: ByteArray(0).toRequestBody()); "DELETE" -> builder.delete(body); else -> builder.get() }
        client.newCall(builder.build()).execute().use { response ->
            val payload = response.body?.bytes() ?: ByteArray(0)
            if (response.code == 401 && retry) { refresh().getOrElse { return@withContext Result.failure(it) }; return@withContext binaryRequest(path, method, body, idempotency, retry = false) }
            if (!response.isSuccessful) return@withContext Result.failure(problem(payload.decodeToString(), response.code))
            Result.success(payload)
        }
    }

    suspend fun refresh(): Result<TokenSet> = refreshMutex.withLock {
        val token = store.refreshToken ?: return@withLock Result.failure(IllegalStateException("Sign in is required."))
        tokenRequest(FormBody.Builder().add("grant_type", "refresh_token").add("client_id", "readypackets-native").add("refresh_token", token).build())
    }

    private suspend fun validAccessToken(): Result<String> {
        store.accessToken?.takeIf { store.accessExpiryEpochMillis > System.currentTimeMillis() }?.let { return Result.success(it) }
        return refresh().map { it.accessToken }
    }

    private suspend fun tokenRequest(body: FormBody): Result<TokenSet> = withContext(Dispatchers.IO) {
        client.newCall(Request.Builder().url("$portal/api/mobile/v1/token").post(body).build()).execute().use { response ->
            val payload = response.body?.string().orEmpty()
            if (!response.isSuccessful) Result.failure(problem(payload, response.code)) else runCatching { json.decodeFromString<TokenSet>(payload) }.onSuccess { store.save(it) }
        }
    }

    private fun problem(payload: String, code: Int): Throwable = runCatching { json.decodeFromString<ApiProblem>(payload) }.map { IllegalStateException(it.detail ?: it.title) }.getOrElse { IllegalStateException("Request failed ($code).") }
    private fun mimeType(file: File): String = when (file.extension.lowercase()) { "m4a" -> "audio/mp4"; "webm" -> "audio/webm"; "mp3" -> "audio/mpeg"; "wav" -> "audio/wav"; "ogg" -> "audio/ogg"; "pdf" -> "application/pdf"; "png" -> "image/png"; "jpg", "jpeg" -> "image/jpeg"; else -> "application/octet-stream" }
}

File: ios/ReadyPackets/ReadyPacketsApp/Core/MobileAPIClient.swift
import Foundation

actor MobileAPIClient {
    private let tokenStore: TokenStore
    private let session: URLSession
    private let decoder: JSONDecoder

    init(tokenStore: TokenStore, session: URLSession = .shared) {
        self.tokenStore = tokenStore
        self.session = session
        decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
    }

    func tokenRequest(form: [String: String]) async throws -> TokenSet {
        var request = URLRequest(url: AppConfig.portalBaseURL.appending(path: "/api/mobile/v1/token"))
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = form.map { "\($0.key.urlEncoded)=\($0.value.urlEncoded)" }.sorted().joined(separator: "&").data(using: .utf8)
        let (data, response) = try await session.data(for: request)
        return try decode(TokenSet.self, data: data, response: response)
    }

    func get<T: Decodable>(_ path: String, as type: T.Type) async throws -> T {
        try await request(path: path, method: "GET", body: nil, idempotencyKey: nil, as: type)
    }

    func post<T: Decodable>(_ path: String, body: Encodable, idempotencyKey: UUID = UUID(), as type: T.Type) async throws -> T {
        try await request(path: path, method: "POST", body: body, idempotencyKey: idempotencyKey.uuidString, as: type)
    }

    func put<T: Decodable>(_ path: String, body: Encodable, idempotencyKey: UUID = UUID(), as type: T.Type) async throws -> T {
        try await request(path: path, method: "PUT", body: body, idempotencyKey: idempotencyKey.uuidString, as: type)
    }

    func delete<T: Decodable>(_ path: String, idempotencyKey: UUID = UUID(), as type: T.Type) async throws -> T {
        try await request(path: path, method: "DELETE", body: nil, idempotencyKey: idempotencyKey.uuidString, as: type)
    }

    func getData(_ path: String) async throws -> Data {
        try await binaryRequest(path: path, method: "GET", body: nil, contentType: nil, idempotencyKey: nil)
    }

    func upload(_ path: String, files: [URL], fields: [String: String], idempotencyKey: UUID = UUID()) async throws -> CustomerFilesPage {
        let boundary = "ReadyPackets-\(UUID().uuidString)"
        var body = Data()
        for (key, value) in fields {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n\(value)\r\n".data(using: .utf8)!)
        }
        for file in files {
            let data = try Data(contentsOf: file)
            let name = file.lastPathComponent.replacingOccurrences(of: "\"", with: "_")
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"files\"; filename=\"\(name)\"\r\n".data(using: .utf8)!)
            body.append("Content-Type: \(mimeType(for: file))\r\n\r\n".data(using: .utf8)!)
            body.append(data); body.append("\r\n".data(using: .utf8)!)
        }
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        let data = try await binaryRequest(path: path, method: "POST", body: body, contentType: "multipart/form-data; boundary=\(boundary)", idempotencyKey: idempotencyKey.uuidString)
        return try decoder.decode(CustomerFilesPage.self, from: data)
    }

    func revoke() async {
        guard let refresh = await tokenStore.refreshToken else { return }
        var request = URLRequest(url: AppConfig.portalBaseURL.appending(path: "/api/mobile/v1/revoke"))
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = "token=\(refresh.urlEncoded)&client_id=readypackets-native".data(using: .utf8)
        _ = try? await session.data(for: request)
    }

    private func request<T: Decodable>(path: String, method: String, body: Encodable?, idempotencyKey: String?, as type: T.Type, attemptedRefresh: Bool = false) async throws -> T {
        let access = try await validAccessToken()
        var request = URLRequest(url: AppConfig.portalBaseURL.appending(path: "/api/mobile/v1" + path))
        request.httpMethod = method
        request.setValue("Bearer \(access)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let idempotencyKey { request.setValue(idempotencyKey, forHTTPHeaderField: "Idempotency-Key") }
        if let body { request.setValue("application/json", forHTTPHeaderField: "Content-Type"); request.httpBody = try JSONEncoder().encode(AnyEncodable(body)) }
        let (data, response) = try await session.data(for: request)
        if let http = response as? HTTPURLResponse, http.statusCode == 401, !attemptedRefresh {
            _ = try await refresh()
            return try await request(path: path, method: method, body: body, idempotencyKey: idempotencyKey, as: type, attemptedRefresh: true)
        }
        return try decode(T.self, data: data, response: response)
    }

    private func binaryRequest(path: String, method: String, body: Data?, contentType: String?, idempotencyKey: String?, attemptedRefresh: Bool = false) async throws -> Data {
        let access = try await validAccessToken()
        var request = URLRequest(url: AppConfig.portalBaseURL.appending(path: "/api/mobile/v1" + path))
        request.httpMethod = method
        request.setValue("Bearer \(access)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let contentType { request.setValue(contentType, forHTTPHeaderField: "Content-Type") }
        if let idempotencyKey { request.setValue(idempotencyKey, forHTTPHeaderField: "Idempotency-Key") }
        request.httpBody = body
        let (data, response) = try await session.data(for: request)
        if let http = response as? HTTPURLResponse, http.statusCode == 401, !attemptedRefresh {
            _ = try await refresh()
            return try await binaryRequest(path: path, method: method, body: body, contentType: contentType, idempotencyKey: idempotencyKey, attemptedRefresh: true)
        }
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else { throw (try? decoder.decode(APIProblem.self, from: data)) ?? APIProblem(title: "Request failed", code: "http_error", detail: nil) }
        return data
    }

    private func validAccessToken() async throws -> String {
        if let token = await tokenStore.accessToken, let expiry = await tokenStore.accessTokenExpiry, expiry > Date() { return token }
        return try await refresh().accessToken
    }

    private func refresh() async throws -> TokenSet {
        guard let refreshToken = await tokenStore.refreshToken else { throw APIProblem(title: "Sign in required", code: "signed_out", detail: "Please sign in again.") }
        let tokens = try await tokenRequest(form: ["grant_type": "refresh_token", "client_id": "readypackets-native", "refresh_token": refreshToken])
        try await tokenStore.save(tokens: tokens)
        return tokens
    }

    private func decode<T: Decodable>(_ type: T.Type, data: Data, response: URLResponse) throws -> T {
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard (200...299).contains(http.statusCode) else { throw (try? decoder.decode(APIProblem.self, from: data)) ?? APIProblem(title: "Request failed", code: "http_\(http.statusCode)", detail: nil) }
        return try decoder.decode(T.self, from: data)
    }

    private func mimeType(for url: URL) -> String {
        switch url.pathExtension.lowercased() {
        case "m4a": return "audio/mp4"
        case "webm": return "audio/webm"
        case "mp3": return "audio/mpeg"
        case "wav": return "audio/wav"
        case "pdf": return "application/pdf"
        case "png": return "image/png"
        case "jpg", "jpeg": return "image/jpeg"
        default: return "application/octet-stream"
        }
    }
}

private struct AnyEncodable: Encodable { let value: Encodable; init(_ value: Encodable) { self.value = value }; func encode(to encoder: Encoder) throws { try value.encode(to: encoder) } }
private extension String { var urlEncoded: String { addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? self } }
